compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2044,17 +2044,34 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                         // If a tuple of length one was expected and the found expression has
                         // parentheses around it, perhaps the user meant to write `(expr,)` to
                         // build a tuple (issue #86100)
-                        (ty::Tuple(_), _) if expected.tuple_fields().count() == 1 => {
-                            if let Ok(code) = self.tcx.sess().source_map().span_to_snippet(span) {
-                                if let Some(code) =
-                                    code.strip_prefix('(').and_then(|s| s.strip_suffix(')'))
-                                {
-                                    err.span_suggestion(
-                                        span,
-                                        "use a trailing comma to create a tuple with one element",
-                                        format!("({},)", code),
-                                        Applicability::MaybeIncorrect,
-                                    );
+                        (ty::Tuple(_), _) => {
+                            if let [expected_tup_elem] =
+                                expected.tuple_fields().collect::<Vec<_>>()[..]
+                            {
+                                if same_type_modulo_infer(expected_tup_elem, found) {
+                                    if let Ok(code) =
+                                        self.tcx.sess().source_map().span_to_snippet(span)
+                                    {
+                                        if code.starts_with('(') && code.ends_with(')') {
```

Comment:
If you want to reduce the nesting, you could extract a helper function like in the sugg-tuple-wrap PR. I'm ok with leaving it as-is if you want though.

Reply:
Ah, I just saw https://github.com/rust-lang/rust/pull/91530#discussion_r794958681. Just because the match arms are local doesn't mean they should be ;)

Reply:
Cool - pulled out to a helper 😄